### PR TITLE
storage: split out a client crate for faster compiles

### PR DESCRIPTION
**Pull request by petrosagg**
### Motivation

Faster iteration cycles.

### Description

Splits out a storage-client crate that is the crate that everything that is not storage should depend on for talking to storage. This allows much faster iteration when working on storaged as touching the code there doesn't trigger a recompilation of environmentd nor computed.

On my machine after this PR touching a file in storage and running `./bin/environmentd` is 3.5x faster.

### Tips for reviewer

This PR is pure code movement

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  -

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -51,7 +51,7 @@ members = [
     "src/stash",
     "src/stash-debug",
     "src/storage",
-    "src/storaged",
```
Comment (guswynn) [inline Cargo.toml:54]: yesss i hated this being different than compute

File: src/sql/src/plan/statement/ddl.rs [hunk 2]
```diff
@@ -1133,15 +1133,14 @@ pub(crate) fn load_generator_ast_to_generator(
     };
 
     let mut available_subsources = BTreeMap::new();
-    let generator = as_generator(&load_generator);
-    for (i, (name, desc)) in generator.views().iter().enumerate() {
+    for (i, (name, desc)) in load_generator.views().iter().enumerate() {
         let name = FullObjectName {
             database: RawDatabaseSpecifier::Name("mz_load_generators".to_owned()),
-            schema: match loadgen {
-                mz_sql_parser::ast::LoadGenerator::Counter => "counter".into(),
```
Comment (guswynn) [inline src/sql/src/plan/statement/ddl.rs:1141]: nice, no longer using the ast types

File: src/storage-client/src/types/sources.rs [hunk 3]
```diff
@@ -1421,7 +1428,7 @@ impl SourceDesc {
 }
 
 #[derive(Arbitrary, Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
-pub enum SourceConnection {
+pub enum GenericSourceConnection {
```
Comment (guswynn) [inline src/storage-client/src/types/sources.rs:1431]: not sure i understand the name change to `GenericSourceConnection`, i think leaving it as is was fine, but i suppose we can iterate on this

**Review by guswynn — APPROVED** [review]
looks like there is a ci failure, but other than that looks good

we probably want to run `bin/unused-deps` at some point after this too!

File: src/storage-client/src/types/sources.rs (hunk 3 above)
Comment (petrosagg) [reply src/storage-client/src/types/sources.rs:1431]: There is a trait with the same name that is now in the same file and it was annoying disambiguating them so I renamed this

**Comment by petrosagg** [comment]
@guswynn thanks for the review! I cleaned up the deps too now :)

**Pull request merged**